### PR TITLE
Adding support for detecting cmp + select as Relu

### PR DESCRIPTION
**Pull request by nhasabni**
Adding unit tests as well and they pass.

Addresses #773.

### CODE REVIEW

**Review by rengolin — COMMENTED** [review]
Great set of tests! You couldn't find a way to use the  op chain method?

File: lib/TPP/IR/MatcherUtils.cpp [hunk 1]
```diff
@@ -151,7 +144,82 @@ static bool hasReluBody(Operation *op, SmallVectorImpl<Value> *captured) {
     }
     return false;
   };
-  return (getOperand(maxfLhs, maxfRhs) || getOperand(maxfRhs, maxfLhs));
+
+  // Multiple patterns map to Relu.
+  if (isa<arith::MaximumFOp>(innerOp)) {
```
Comment (chelini) [inline lib/TPP/IR/MatcherUtils.cpp:149]: I would change as: `if (auto maxfOp = dyn_cast<arith::MaximumFOp>(innerOp))`. This allows you yo avoid the cast in line 153.

File: lib/TPP/IR/MatcherUtils.cpp [hunk 2]
```diff
@@ -151,7 +144,82 @@ static bool hasReluBody(Operation *op, SmallVectorImpl<Value> *captured) {
...
+    // Pattern 1 - arith.maximumf(tensor, const 0)
+    if (yieldOp->getOperand(0).getDefiningOp() != innerOp)
+      return false;
+    auto maxfOp = cast<arith::MaximumFOp>(innerOp);
+    Value maxfLhs = maxfOp.getLhs();
+    Value maxfRhs = maxfOp.getRhs();
+
+    return (getOperand(maxfLhs, maxfRhs) || getOperand(maxfRhs, maxfLhs));
+  } else if (isa<arith::CmpFOp>(innerOp)) {
```
Comment (chelini) [inline lib/TPP/IR/MatcherUtils.cpp:158]: same here.

File: lib/TPP/IR/MatcherUtils.cpp [hunk 3]
```diff
@@ -151,7 +144,82 @@ static bool hasReluBody(Operation *op, SmallVectorImpl<Value> *captured) {
...
+    // Pattern 2 - arith.cmpf, arith.select
+    // %22 = arith.cmpf ugt, %in, %cst : f32
```
Comment (chelini) [inline lib/TPP/IR/MatcherUtils.cpp:160]: I would drop this IR.

**Review by chelini — COMMENTED** [review]
thanks, only minor comments for now.

File: lib/TPP/IR/MatcherUtils.cpp (hunk 3 above)
Comment (nhasabni) [reply lib/TPP/IR/MatcherUtils.cpp:160]: I was keeping it as a doc for easier code understanding.

Comment (nhasabni) [reply lib/TPP/IR/MatcherUtils.cpp]: Reworded it to make it more abstract.

**Comment by nhasabni** [comment]
Thanks @chelini @rengolin  for comments. Please take a look after the change now.

**Comment by rengolin** [comment]
> Great set of tests! You couldn't find a way to use the op chain method?

Discussed offline with @nhasabni, we'll defer the move to structured matchers because the patterns here are not that simple and we may need some new support in the matchers themselves. For now, we'll carry on with the bloated function and refactor later.

**Review by rengolin — APPROVED** [review]
Sorry it took so long, it fell out of my radar. LGTM, thanks!

**Pull request merged**